flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/TableFunction.scala

```diff
@@ -135,8 +135,11 @@ abstract class TableFunction[T] extends UserDefinedFunction {
     *
     * @param arguments arguments of a function call (only literal arguments
     *                  are passed, nulls for non-literal ones)
+    * @param typeInfos The type information of the parameters. only valid argument types
+    *                  are passed, nulls for non-valid ones.
     * @return [[TypeInformation]] of result type or null if Flink should determine the type
     */
-  def getResultType(arguments: java.util.List[AnyRef]): TypeInformation[T] = null
+  def getResultType(arguments: java.util.List[AnyRef],
+                    typeInfos: java.util.List[TypeInformation[_]]): TypeInformation[T] = null
```

Comment:
I was thinking about this again. Maybe we should pass a `List[Class[_]]` rather than `List[TypeInformation[_]]`. The `eval` method receives regular classes (or primitives) and does not have to deal with type informations.
What do you think, @clarkyzl and @KurtYoung?